File: android/src/main/kotlin/dev/enb/vidsqueeze/VidsqueezePlugin.kt
package dev.enb.vidsqueeze

import android.content.Context
import android.net.Uri
import dev.enb.compressi.core.CompressionFailure
import dev.enb.compressi.core.CompressionHandle
import dev.enb.compressi.core.CompressionListener
import dev.enb.compressi.core.CompressionPreset
import dev.enb.compressi.core.CompressionRequest
import dev.enb.compressi.core.CompressionState
import dev.enb.compressi.core.CompressionSuccess
import dev.enb.compressi.core.ForceCodec
import dev.enb.compressi.core.VideoCompressor
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.UUID
import java.util.concurrent.atomic.AtomicReference

class VidsqueezePlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private var applicationContext: Context? = null
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private val activeTaskId = AtomicReference<String?>(null)
    private var compressor: VideoCompressor? = null
    private var activeHandle: CompressionHandle? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        compressor = VideoCompressor(binding.applicationContext)
        methodChannel = MethodChannel(binding.binaryMessenger, FlutterContract.METHODS_CHANNEL).also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, FlutterContract.EVENTS_CHANNEL).also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        activeHandle?.cancel()
        compressor?.shutdown()
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        methodChannel = null
        eventChannel = null
        applicationContext = null
        compressor = null
        eventSink = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            FlutterContract.METHOD_COMPRESS -> compress(call, result)
            FlutterContract.METHOD_CANCEL -> cancel(call, result)
            else -> result.notImplemented()
        }
    }

    private fun compress(call: MethodCall, result: MethodChannel.Result) {
        if (activeHandle != null) {
            result.error(
                FlutterContract.ERROR_BUSY,
                "Another compression task is already running",
                null,
            )
            return
        }

        applicationContext ?: run {
            result.error(
                FlutterContract.ERROR_NO_CONTEXT,
                "Plugin is not attached to an Android context",
                null,
            )
            return
        }
        val compressor = compressor ?: run {
            result.error(
                FlutterContract.ERROR_NO_COMPRESSOR,
                "Video compressor is unavailable",
                null,
            )
            return
        }

        val args = call.arguments as? Map<*, *> ?: run {
            result.error(
                FlutterContract.ERROR_BAD_ARGS,
                "compress expects a map payload",
                null,
            )
            return
        }

        val taskId = args.stringOrNull(FlutterContract.KEY_TASK_ID)
            ?.takeIf { it.isNotBlank() }
            ?: UUID.randomUUID().toString()
        val request = runCatching { args.toCompressionRequest(taskId) }.getOrElse { throwable ->
            val failure = throwable as? CompressionFailure
            result.error(
                failure?.code?.name?.lowercase() ?: FlutterContract.ERROR_BAD_REQUEST,
                throwable.message,
                mapOf(FlutterContract.KEY_TASK_ID to taskId),
            )
            return
        }

        activeTaskId.set(taskId)
        activeHandle = compressor.start(
            request,
            object : CompressionListener {
                override fun onStateChanged(state: CompressionState) {
                    eventSink?.success(state.toMap(taskId))
                }

                override fun onSuccess(resultValue: CompressionSuccess) {
                    activeHandle = null
                    activeTaskId.set(null)
                    result.success(resultValue.toMap(taskId))
                }

                override fun onFailure(failure: CompressionFailure) {
                    activeHandle = null
                    activeTaskId.set(null)
                    result.error(
                        failure.code.name.lowercase(),
                        failure.message,
                        mapOf(FlutterContract.KEY_TASK_ID to taskId),
                    )
                }
            },
        )
    }

    private fun cancel(call: MethodCall, result: MethodChannel.Result) {
        val taskId = (call.arguments as? Map<*, *>)?.stringOrNull(FlutterContract.KEY_TASK_ID)
        if (taskId == null || taskId == activeTaskId.get()) {
            activeHandle?.cancel()
            activeHandle = null
            activeTaskId.set(null)
        }
        result.success(null)
    }
}

private fun Map<*, *>.toCompressionRequest(taskId: String): CompressionRequest {
    val inputPath = stringOrNull(FlutterContract.KEY_INPUT_PATH)
        ?: error("${FlutterContract.KEY_INPUT_PATH} is required")
    val outputDirectoryPath = stringOrNull(FlutterContract.KEY_OUTPUT_DIRECTORY_PATH)
        ?: error("${FlutterContract.KEY_OUTPUT_DIRECTORY_PATH} is required")
    val outputFileName = stringOrNull(FlutterContract.KEY_OUTPUT_FILE_NAME)
        ?.takeIf { it.isNotBlank() }
        ?: "compressed_${taskId}.mp4"
    val preset = CompressionPreset.valueOf(
        (stringOrNull(FlutterContract.KEY_PRESET) ?: "balanced").toCompressionPresetName(),
    )
    val forceCodec = ForceCodec.valueOf(
        (stringOrNull(FlutterContract.KEY_FORCE_CODEC) ?: "auto").uppercase(),
    )

    return CompressionRequest(
        inputUri = Uri.parse(inputPath),
        outputDirectory = outputDirectoryPath.toPlatformFile(),
        outputFileName = outputFileName,
        preset = preset,
        maxResolutionCap = (get(FlutterContract.KEY_MAX_RESOLUTION_CAP) as? Number)?.toInt(),
        allowHevc = get(FlutterContract.KEY_ALLOW_HEVC) as? Boolean ?: true,
        keepAudio = get(FlutterContract.KEY_KEEP_AUDIO) as? Boolean ?: true,
        keepOriginalIfLarger = get(FlutterContract.KEY_KEEP_ORIGINAL_IF_LARGER) as? Boolean ?: true,
        forceCodec = forceCodec,
        maxBitrate = (get(FlutterContract.KEY_MAX_BITRATE) as? Number)?.toInt(),
        progressIntervalMs = (get(FlutterContract.KEY_PROGRESS_INTERVAL_MS) as? Number)?.toLong() ?: 250L,
    )
}

private fun String.toCompressionPresetName(): String {
    return when (lowercase()) {
        "quality" -> "QUALITY"
        "small_size" -> "SMALL_SIZE"
        else -> "BALANCED"
    }
}

private fun CompressionState.toMap(taskId: String): Map<String, Any?> {
    return when (this) {
        CompressionState.Preparing -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_PREPARING,
        )
        CompressionState.Finalizing -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_FINALIZING,
        )
        CompressionState.Completed -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_COMPLETED,
        )
        CompressionState.Cancelled -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_CANCELLED,
        )
        is CompressionState.Transcoding -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_TRANSCODING,
            FlutterContract.KEY_PROGRESS_PERCENT to progressPercent,
        )
        is CompressionState.Failed -> mapOf(
            FlutterContract.KEY_TASK_ID to taskId,
            FlutterContract.KEY_PHASE to FlutterContract.PHASE_FAILED,
            FlutterContract.KEY_CODE to failure.code.name.lowercase(),
            FlutterContract.KEY_MESSAGE to failure.message,
        )
    }
}

private fun CompressionSuccess.toMap(taskId: String): Map<String, Any?> {
    return mapOf(
        FlutterContract.KEY_TASK_ID to taskId,
        FlutterContract.KEY_OUTPUT_PATH to outputFile.absolutePath,
        FlutterContract.KEY_OUTPUT_SIZE_BYTES to outputSizeBytes,
        FlutterContract.KEY_SOURCE_SIZE_BYTES to sourceSizeBytes,
        FlutterContract.KEY_DURATION_MS to durationMs,
        FlutterContract.KEY_CODEC to codec.name.lowercase(),
        FlutterContract.KEY_TARGET_HEIGHT to targetHeight,
        FlutterContract.KEY_TARGET_BITRATE to targetBitrate,
        FlutterContract.KEY_ATTEMPTS to attempts,
        FlutterContract.KEY_USED_ORIGINAL_SOURCE to usedOriginalSource,
    )
}

private fun Map<*, *>.stringOrNull(key: String): String? = get(key) as? String

private fun String.toPlatformFile(): File {
    return if (startsWith("file://")) {
        File(requireNotNull(Uri.parse(this).path) { "Invalid file URI: $this" })
    } else {
        File(this)
    }
}

private object FlutterContract {
    const val METHODS_CHANNEL = "vidsqueeze/methods"
    const val EVENTS_CHANNEL = "vidsqueeze/events"

    const val METHOD_COMPRESS = "compress"
    const val METHOD_CANCEL = "cancel"

    const val KEY_TASK_ID = "taskId"
    const val KEY_INPUT_PATH = "inputPath"
    const val KEY_OUTPUT_DIRECTORY_PATH = "outputDirectoryPath"
    const val KEY_OUTPUT_FILE_NAME = "outputFileName"
    const val KEY_PRESET = "preset"
    const val KEY_MAX_RESOLUTION_CAP = "maxResolutionCap"
    const val KEY_ALLOW_HEVC = "allowHevc"
    const val KEY_KEEP_AUDIO = "keepAudio"
    const val KEY_KEEP_ORIGINAL_IF_LARGER = "keepOriginalIfLarger"
    const val KEY_FORCE_CODEC = "forceCodec"
    const val KEY_MAX_BITRATE = "maxBitrate"
    const val KEY_PROGRESS_INTERVAL_MS = "progressIntervalMs"
    const val KEY_PHASE = "phase"
    const val KEY_PROGRESS_PERCENT = "progressPercent"
    const val KEY_CODE = "code"
    const val KEY_MESSAGE = "message"
    const val KEY_OUTPUT_PATH = "outputPath"
    const val KEY_OUTPUT_SIZE_BYTES = "outputSizeBytes"
    const val KEY_SOURCE_SIZE_BYTES = "sourceSizeBytes"
    const val KEY_DURATION_MS = "durationMs"
    const val KEY_CODEC = "codec"
    const val KEY_TARGET_HEIGHT = "targetHeight"
    const val KEY_TARGET_BITRATE = "targetBitrate"
    const val KEY_ATTEMPTS = "attempts"
    const val KEY_USED_ORIGINAL_SOURCE = "usedOriginalSource"

    const val PHASE_PREPARING = "preparing"
    const val PHASE_TRANSCODING = "transcoding"
    const val PHASE_FINALIZING = "finalizing"
    const val PHASE_COMPLETED = "completed"
    const val PHASE_FAILED = "failed"
    const val PHASE_CANCELLED = "cancelled"

    const val ERROR_BUSY = "busy"
    const val ERROR_BAD_ARGS = "bad_args"
    const val ERROR_BAD_REQUEST = "bad_request"
    const val ERROR_NO_CONTEXT = "no_context"
    const val ERROR_NO_COMPRESSOR = "no_compressor"
}

File: ios/Classes/VidsqueezePlugin.swift
#if canImport(UIKit) && canImport(Flutter)
import Flutter
import Foundation
import UIKit

public final class VidsqueezePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private let compressor = VidsqueezeVideoCompressor()
    private var eventSink: FlutterEventSink?
    private var activeTaskId: String?
    private var activeHandle: VidsqueezeCompressionHandle?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VidsqueezePlugin()
        let methodChannel = FlutterMethodChannel(
            name: FlutterContract.methodsChannel,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: FlutterContract.eventsChannel,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case FlutterContract.methodCompress:
            compress(call: call, result: result)
        case FlutterContract.methodCancel:
            cancel(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func compress(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard activeHandle == nil else {
            result(
                FlutterError(
                    code: FlutterContract.errorBusy,
                    message: "Another compression task is already running",
                    details: nil
                )
            )
            return
        }
        guard let args = call.arguments as? [String: Any] else {
            result(
                FlutterError(
                    code: FlutterContract.errorBadArgs,
                    message: "compress expects a map payload",
                    details: nil
                )
            )
            return
        }

        let taskId = args.stringValue(for: FlutterContract.keyTaskId)?
            .nonEmpty
            ?? UUID().uuidString

        do {
            let request = try args.toCompressionRequest(taskId: taskId)
            let listener = PluginCompressionListener(taskId: taskId, result: result) { [weak self] payload in
                self?.eventSink?(payload)
            } onComplete: { [weak self] in
                self?.activeHandle = nil
                self?.activeTaskId = nil
            }

            activeTaskId = taskId
            activeHandle = compressor.start(request: request, listener: listener)
        } catch let failure as VidsqueezeCompressionFailure {
            result(
                FlutterError(
                    code: failure.code.rawValue,
                    message: failure.message,
                    details: [FlutterContract.keyTaskId: taskId]
                )
            )
        } catch {
            result(
                FlutterError(
                    code: FlutterContract.errorBadRequest,
                    message: error.localizedDescription,
                    details: [FlutterContract.keyTaskId: taskId]
                )
            )
        }
    }

    private func cancel(call: FlutterMethodCall, result: FlutterResult) {
        let taskId = (call.arguments as? [String: Any])?.stringValue(for: FlutterContract.keyTaskId)
        if taskId == nil || taskId == activeTaskId {
            activeHandle?.cancel()
            activeHandle = nil
            activeTaskId = nil
        }
        result(nil)
    }
}

private final class PluginCompressionListener: NSObject, VidsqueezeCompressionListener, @unchecked Sendable {
    private let taskId: String
    private let result: FlutterResult
    private let emitState: ([String: Any?]) -> Void
    private let onComplete: () -> Void
    private var didComplete = false

    init(
        taskId: String,
        result: @escaping FlutterResult,
        emitState: @escaping ([String: Any?]) -> Void,
        onComplete: @escaping () -> Void
    ) {
        self.taskId = taskId
        self.result = result
        self.emitState = emitState
        self.onComplete = onComplete
    }

    func onStateChanged(_ state: VidsqueezeCompressionState) {
        emitState(state.toMap(taskId: taskId))
    }

    func onSuccess(_ resultValue: VidsqueezeCompressionSuccess) {
        guard !didComplete else { return }
        didComplete = true
        onComplete()
        result(resultValue.toMap(taskId: taskId))
    }

    func onFailure(_ failure: VidsqueezeCompressionFailure) {
        guard !didComplete else { return }
        didComplete = true
        onComplete()
        result(
            FlutterError(
                code: failure.code.rawValue,
                message: failure.message,
                details: [FlutterContract.keyTaskId: taskId]
            )
        )
    }
}

private extension Dictionary where Key == String, Value == Any {
    func toCompressionRequest(taskId: String) throws -> VidsqueezeCompressionRequest {
        guard let inputPath = stringValue(for: FlutterContract.keyInputPath) else {
            throw VidsqueezeCompressionFailure(
                code: .unsupportedInput,
                message: "\(FlutterContract.keyInputPath) is required"
            )
        }
        guard let outputDirectoryPath = stringValue(for: FlutterContract.keyOutputDirectoryPath) else {
            throw VidsqueezeCompressionFailure(
                code: .unsupportedInput,
                message: "\(FlutterContract.keyOutputDirectoryPath) is required"
            )
        }

        let outputFileName = stringValue(for: FlutterContract.keyOutputFileName)?.nonEmpty
            ?? "compressed_\(taskId).mp4"
        let inputURL = inputPath.toPlatformURL()
        let outputURL = outputDirectoryPath
            .toPlatformURL(isDirectory: true)
            .appendingPathComponent(outputFileName)

        return try VidsqueezeCompressionRequest(
            inputURL: inputURL,
            outputURL: outputURL,
            preset: VidsqueezeCompressionPreset(rawValue: stringValue(for: FlutterContract.keyPreset) ?? "balanced") ?? .balanced,
            maxResolutionCap: (self[FlutterContract.keyMaxResolutionCap] as? NSNumber)?.intValue,
            allowHevc: self[FlutterContract.keyAllowHevc] as? Bool ?? true,
            keepAudio: self[FlutterContract.keyKeepAudio] as? Bool ?? true,
            keepOriginalIfLarger: self[FlutterContract.keyKeepOriginalIfLarger] as? Bool ?? true,
            forceCodec: VidsqueezeForceCodec(rawValue: stringValue(for: FlutterContract.keyForceCodec) ?? "auto") ?? .auto,
            maxBitrate: (self[FlutterContract.keyMaxBitrate] as? NSNumber)?.intValue,
            progressIntervalMs: (self[FlutterContract.keyProgressIntervalMs] as? NSNumber)?.intValue ?? 250
        )
    }

    func stringValue(for key: String) -> String? {
        self[key] as? String
    }
}

private extension VidsqueezeCompressionState {
    func toMap(taskId: String) -> [String: Any?] {
        switch self {
        case .preparing:
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phasePreparing,
            ]
        case .finalizing:
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phaseFinalizing,
            ]
        case .completed:
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phaseCompleted,
            ]
        case .cancelled:
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phaseCancelled,
            ]
        case let .transcoding(progressPercent):
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phaseTranscoding,
                FlutterContract.keyProgressPercent: progressPercent,
            ]
        case let .failed(code, message):
            return [
                FlutterContract.keyTaskId: taskId,
                FlutterContract.keyPhase: FlutterContract.phaseFailed,
                FlutterContract.keyCode: code.rawValue,
                FlutterContract.keyMessage: message,
            ]
        }
    }
}

private extension VidsqueezeCompressionSuccess {
    func toMap(taskId: String) -> [String: Any?] {
        [
            FlutterContract.keyTaskId: taskId,
            FlutterContract.keyOutputPath: outputURL.path,
            FlutterContract.keyOutputSizeBytes: outputSizeBytes,
            FlutterContract.keySourceSizeBytes: sourceSizeBytes,
            FlutterContract.keyDurationMs: durationMs,
            FlutterContract.keyCodec: codec.rawValue,
            FlutterContract.keyTargetHeight: targetHeight,
            FlutterContract.keyTargetBitrate: targetBitrate,
            FlutterContract.keyAttempts: attempts,
            FlutterContract.keyUsedOriginalSource: usedOriginalSource,
        ]
    }
}

private extension String {
    var nonEmpty: String? {
        isEmpty ? nil : self
    }

    func toPlatformURL(isDirectory: Bool = false) -> URL {
        if hasPrefix("file://"), let url = URL(string: self) {
            return url
        }
        return URL(fileURLWithPath: self, isDirectory: isDirectory)
    }
}

private enum FlutterContract {
    static let methodsChannel = "vidsqueeze/methods"
    static let eventsChannel = "vidsqueeze/events"

    static let methodCompress = "compress"
    static let methodCancel = "cancel"

    static let keyTaskId = "taskId"
    static let keyInputPath = "inputPath"
    static let keyOutputDirectoryPath = "outputDirectoryPath"
    static let keyOutputFileName = "outputFileName"
    static let keyPreset = "preset"
    static let keyMaxResolutionCap = "maxResolutionCap"
    static let keyAllowHevc = "allowHevc"
    static let keyKeepAudio = "keepAudio"
    static let keyKeepOriginalIfLarger = "keepOriginalIfLarger"
    static let keyForceCodec = "forceCodec"
    static let keyMaxBitrate = "maxBitrate"
    static let keyProgressIntervalMs = "progressIntervalMs"
    static let keyPhase = "phase"
    static let keyProgressPercent = "progressPercent"
    static let keyCode = "code"
    static let keyMessage = "message"
    static let keyOutputPath = "outputPath"
    static let keyOutputSizeBytes = "outputSizeBytes"
    static let keySourceSizeBytes = "sourceSizeBytes"
    static let keyDurationMs = "durationMs"
    static let keyCodec = "codec"
    static let keyTargetHeight = "targetHeight"
    static let keyTargetBitrate = "targetBitrate"
    static let keyAttempts = "attempts"
    static let keyUsedOriginalSource = "usedOriginalSource"

    static let phasePreparing = "preparing"
    static let phaseTranscoding = "transcoding"
    static let phaseFinalizing = "finalizing"
    static let phaseCompleted = "completed"
    static let phaseFailed = "failed"
    static let phaseCancelled = "cancelled"

    static let errorBusy = "busy"
    static let errorBadArgs = "bad_args"
    static let errorBadRequest = "bad_request"
}
#endif
